text/0000-dyn-error-generic-member-access.md

```diff
@@ -0,0 +1,527 @@
+}
+```
+
+### Use this `Request` type to handle passing generic types out of the trait object
+
+```rust
+impl dyn Error {
+    pub fn context<T: ?Sized + 'static>(&self) -> Option<&T> {
+        Request::with::<T, _>(|req| self.provide_context(req))
+    }
+}
+```
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+* The `Request` api is being added purely to help with this function. This will
+  likely need some design work to make it more generally applicable, hopefully
+  as a struct in `core::any`.
+* The `context` function name is currently widely used throughout the rust
+  error handling ecosystem in libraries like `anyhow` and `snafu` as an
+  ergonomic version of `map_err`. If we settle on `context` as the final name
+  it will possibly break existing libraries.
+
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+## Do Nothing
+
+We could not do this, and continue to add accessor functions to the `Error`
+trait whenever a new type reaches critical levels of popularity in error
+reporting.
+
+If we choose to do nothing we will continue to see hacks around the current
+limitations on the `Error` trait such as the `Fail` trait, which added the
+missing function access methods that didn't previously exist on the `Error`
+trait and type erasure / unnecessary boxing of errors to enable downcasting to
+extract members.
+[[1]](https://docs.rs/tracing-error/0.1.2/src/tracing_error/error.rs.html#269-274).
+
+
+## Use an alternative proposal that relies on the `Any` trait for downcasting
+
+```rust
+pub trait Error {
+    /// Provide an untyped reference to a member whose type matches the provided `TypeId`.
+    ///
+    /// Returns `None` by default, implementors are encouraged to override.
+    fn provide_context(&self, ty: TypeId) -> Option<&dyn Any> {
+        None
+    }
+}
+
+impl dyn Error {
+    /// Retrieve a reference to `T`-typed context from the error if it is available.
+    pub fn context<T: Any>(&self) -> Option<&T> {
+        self.provide_context(TypeId::of::<T>())?.downcast_ref::<T>()
+    }
+}
+```
+
+### Why isn't this the primary proposal?
+
+There are two big issues with using the `Any` trait that I believe justify the
+more complicated solution.
+
+- You cannot return dynamically sized types as `&dyn Any`
+- It's easy to introduce runtime errors with `&dyn Any` by either comparing to
+  or returning the wrong type
+
+By making all the type id comparison internal to the `Request` type it is
+impossible to compare the wrong type ids. By encouraging explicit type
+parameters when calling `provide` the compiler is able to catch errors where
+the type passed in doesn't match the type that was expected. So while the API
+for the main proposal is more complicated it should be less error prone.
+
+# Prior art
+[prior-art]: #prior-art
+
+I do not know of any other languages whose error handling has similar
+facilities for accessing members when reporting errors. For the most part,
```

Comment:
AIUI the analogous operation in a class-inheritance language like Java would be to attempt a cast to a subclass, or am I missing something? This seems like the mix-in version of that if I squint.

Reply:
yea. This could probably be used for upcasting too, you just have to return `&self`, I'm not really sure if thats prior art but ill see about mentioning similarities.

Reply:
Yeah it’s more “prior niche fulfillment” than “prior art”.